pkg/controllers/management/clusterprovisioner/provisioner.go

```diff
@@ -214,6 +245,39 @@ func (p *Provisioner) Updated(cluster *v3.Cluster) (runtime.Object, error) {
 	return obj.(*v3.Cluster), err
 }
 
+func (p *Provisioner) startCleanUpHelmReleases(cluster *v3.Cluster) error {
+	err := p.cleanupHelmReleases(cluster)
+	if err != nil {
+		return err
+	}
+
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		cluster, err = p.Clusters.Get(cluster.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// Attempt to manually trigger updating, otherwise it will not be triggered until after exiting reconcile
+		apimgmtv3.ClusterConditionHelmReleasesMigrated.True(cluster)
+		apimgmtv3.ClusterConditionHelmReleasesMigrated.Message(cluster, "Helm releases cleaned up")
+
+		cluster, err = p.Clusters.Update(cluster)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
```

Comment:
You can wrap the logic above this in <condition>. Do and True will be applied to the cluster on success and an error will set it to False. Does a conflict happen often error return often? I don't like using these kinds of retry loops unless there is a sizable efficiency payoff because the handler framework itself will handle retries when an error is returned. Your logic seems like it will know what to do on re enqeue since it will retry until the condition is set and since the release logic was done it will be shorter next time.

Reply:
This change was requested here: https://github.com/rancher/rancher/pull/40104#discussion_r1068268588

Reply:
Updated!